torch/optim/_multi_tensor/sgd.py

```diff
@@ -0,0 +1,140 @@
+
+        Considering the specific case of Momentum, the update can be written as
+
+        .. math::
+            \begin{aligned}
+                v_{t+1} & = \mu * v_{t} + g_{t+1}, \\
+                p_{t+1} & = p_{t} - \text{lr} * v_{t+1},
+            \end{aligned}
+
+        where :math:`p`, :math:`g`, :math:`v` and :math:`\mu` denote the 
+        parameters, gradient, velocity, and momentum respectively.
+
+        This is in contrast to Sutskever et. al. and
+        other frameworks which employ an update of the form
+
+        .. math::
+            \begin{aligned}
+                v_{t+1} & = \mu * v_{t} + \text{lr} * g_{t+1}, \\
+                p_{t+1} & = p_{t} - v_{t+1}.
+            \end{aligned}
+
+        The Nesterov version is analogously modified.
+    """
+
+    def __init__(self, params, lr=required, momentum=0, dampening=0,
+                 weight_decay=0, nesterov=False):
+        if lr is not required and lr < 0.0:
+            raise ValueError("Invalid learning rate: {}".format(lr))
+        if momentum < 0.0:
+            raise ValueError("Invalid momentum value: {}".format(momentum))
+        if weight_decay < 0.0:
+            raise ValueError("Invalid weight_decay value: {}".format(weight_decay))
+
+        defaults = dict(lr=lr, momentum=momentum, dampening=dampening,
+                        weight_decay=weight_decay, nesterov=nesterov)
+        if nesterov and (momentum <= 0 or dampening != 0):
+            raise ValueError("Nesterov momentum requires a momentum and zero dampening")
+        super(SGD, self).__init__(params, defaults)
+
+    def __setstate__(self, state):
+        super(SGD, self).__setstate__(state)
+        for group in self.param_groups:
+            group.setdefault('nesterov', False)
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        """Performs a single optimization step.
+
+        Arguments:
+            closure (callable, optional): A closure that reevaluates the model
+                and returns the loss.
+        """
+        loss = None
+        if closure is not None:
+            with torch.enable_grad():
+                loss = closure()
+
+        for group in self.param_groups:
+            weight_decay = group['weight_decay']
+            momentum = group['momentum']
+            dampening = group['dampening']
+            nesterov = group['nesterov']
+
+            grads = []
+            params_with_grad = []
+            states = []
+            has_sparse_grad = False
+
+            for p in group['params']:
+                if p.grad is not None:
+                    grads.append(p.grad)
+                    params_with_grad.append(p)
+                    states.append(self.state[p])
+
+                    if p.grad.is_sparse:
+                        has_sparse_grad = True
+
+                        if momentum != 0: 
+                            raise RuntimeError('SGD does not support momentum for sparse gradients')
```

Comment:
The CPU overhead of this region is substantial (corresponds to the `"cpu packaging"` nvtx range of the attached profile).

Profile is for resnet50+native amp.  There's an amp-related sync (CPU waits for the GPU) just before SGD.step(), so the CPU latency is fully exposed to the GPU and causes a bubble on the GPU timeline.  For profiling purposes that's a good thing, because it clearly shows how much CPU overhead we have.  In a non-amp run, the overhead might not cause a GPU bubble because there's no sync, so the CPU might have enough time to get the packaging done and the optimizer kernels enqueued while the GPU is still processing earlier work.  (In amp-enabled runs, in principle we can similarly reduce the risk of a GPU bubble with "sync-free dynamic loss scaling," an extra layer of complexity we should consider soon, but not for 1.7.)

With or without an amp sync (and whether or not it causes a bubble on the GPU) there's a big chunk of CPU overhead here.  Some overhead is inevitable, but if there's time tomorrow, we should consider if any Python tricks can moderately reduce the overhead.
![mta_sgd_step](https://user-images.githubusercontent.com/7799218/94395880-e0ac4e00-011d-11eb-99c7-004d24096694.png)